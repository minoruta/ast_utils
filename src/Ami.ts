/**
 * A utility library for asterisk
 */
import * as AmiClient from 'asterisk-ami-client';
import * as DEBUG from 'debug';

const debug = DEBUG('AST_MONGO:ami');

/** configuration to make Ari */
export interface AmiConfigs {
    user: string;
    secret: string;
    host: string;
    port: number;
}

export interface AmiResponse {
  Message: string;
  Output: string;
  Response: string;
}

class Timeout {
    private static readonly DEFAULT_TIMEOUT = 100 * 1000; // msec
    private readonly _saved: number;
    constructor(timeout = Timeout.DEFAULT_TIMEOUT) {
        this._saved = jasmine.DEFAULT_TIMEOUT_INTERVAL;
        jasmine.DEFAULT_TIMEOUT_INTERVAL = timeout;
    }
    end(): void {
        jasmine.DEFAULT_TIMEOUT_INTERVAL = this._saved;
    }
}

/**
 * Ami provides some utility by Asterisk Management Interface.
 */
export class Ami extends AmiClient {

    private _configs: AmiConfigs;

    /**
     * @param configs    is configuration to make Ami
     */
    constructor(configs: AmiConfigs) {
        super({
            reconnect: true,
            keepAlive: true
        });
        this._configs = configs;
    }

    protected get user(): string {
        return this._configs.user;
    }

    protected get secret(): string {
        return this._configs.secret;
    }

    protected get dst(): { host: string, port: number } {
        return {
            host: this._configs.host,
            port: this._configs.port
        };
    }

    /**
     * connect a remort asterisk
     */
    connect(): Promise<void> {
        debug(`connect to ${this.user}@${this.dst.host}:${this.dst.port}`);
        return super.connect(this.user, this.secret, this.dst);
    }

    /**
     * disconnect from the remote server
     */
    disconnect(): void {
        debug('disconnect');
        super.disconnect();
    }

    exec(command: string): Promise<AmiResponse> {
        debug(`exec '${command}'`);
        return super.action({
            Action: 'Command',
            Command: command
        }, true);
    }

    /**
     * let a remote asterisk reload dial plan
     */
    reloadDialPlan(): Promise<AmiResponse> {
        return this.exec('dialplan reload');
    }

    /**
     * execute 'core reload'
     * @return after the command has been executed.
     */
    reload(): Promise<AmiResponse> {
        return this.exec('core reload');
    }

    /**
     * let a remote asterisk-self restart
     * @param when  ['now'|'gracefully'|'when convenient '], the default is 'now'
     * @return when the connection has been reestablished.
     */
    restart(when = 'now'): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            debug('restart');
            const timer = new Timeout();
            super.once('reconnection', () => {
                debug('reconnected');
                timer.end();
                resolve();
            });
            this.exec(`core restart ${when}`)
            .then(response => debug(response))
            .catch((err: any) => {
                if (err.message !== 'Client disconnected.') {
                    timer.end();
                    super.removeAllListeners();
                    return reject(err);
                }
                debug('disconnected');
            });
        });
    }

    /**
     * let a remote server call to callee with specified sound
     * @param callee    id of callee
     * @param sound     sound file to play while calling
     * @param tech      default is 'PJSIP'
     */
    callWithSound(callee: string, sound: string, tech = 'PJSIP'): Promise<AmiResponse> {
        return this.exec(`channel originate ${tech}/${callee} application playback ${sound}`);
    }
}
