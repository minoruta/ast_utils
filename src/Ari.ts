/**
 * A utility library for asterisk
 */
import * as ARI from 'ari-client';
import * as DEBUG from 'debug';

const debug = DEBUG('AST_MONGO:ari');

/** configuration to make Ari */
export interface AriConfigs {
    url: string;
    user: string;
    password: string;
}

/**
 * Ari provides some utility by Asterisk Restful Interface.
 */
export class Ari {

    private _configs: AriConfigs;
    private _ari?: any;

    constructor(configs: AriConfigs) {
        this._configs = configs;
    }

    protected get url(): string {
        return this._configs.url;
    }

    protected get user(): string {
        return this._configs.user;
    }

    protected get password(): string {
        return this._configs.password;
    }

    protected get ari() {
        if (!this._ari)
            throw new Error('not connected');
        return this._ari;
    }

    /**
     * connect a remort asterisk
     */
    async connect(): Promise<void> {
        debug(`connect to ${this.user}@${this.url}`);
        this._ari = await ARI.connect(this.url, this.user, this.password);
    }

    /**
     * disconnect from the remort asterisk
     */
    disconnect(): void {
        debug('disconnect');
        // this.ari.stop();
    }

    /**
     * get an endpoint status of the remote asterisk
     * @param id    is id for the endpoint you want
     */
    async getEndpointStatus(id: string, tech = 'PJSIP'): Promise<string> {
        let state: string;
        try {
            const endpoint = await this.ari.endpoints.get({
                resource: id,
                tech: tech,
            });
            state = endpoint.state;
        }
        catch (err) {
            if (!err.message)
                throw err;
            const message = JSON.parse(err.message);
            if (message.message != 'Endpoint not found')
                throw err;
            state = 'no endpoint';
        }
        debug(`status of ${tech}/${id} is '${state}'`);
        return state;
    }
}
