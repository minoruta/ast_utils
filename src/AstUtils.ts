/**
 * A utility library for asterisk
 */
import { Ari, AriConfigs } from './Ari';
import { Ami, AmiConfigs, AmiResponse } from './Ami';
import * as DEBUG from 'debug';

const debug = DEBUG('AST_MONGO:utils');

/** configuration to make AstUtils */
export interface AstUtilsConfg {
    host: string;
    ari: {
        protocol: string;
        port: number;
        username: string;
        password: string;
    };
    ami: {
        port: number;
        username: string;
        password: string;
    };
}

export type AstUtilsResponse = AmiResponse;

/**
 * AstUtils provides convenient functions to test ast_mongo.
 */
export class AstUtils {

    private _ari?: Ari;
    private _ami?: Ami;

    protected get ari(): Ari {
        if (!this._ari)
            throw new Error('no valid ari');
        return this._ari;
    }

    protected get ami(): Ami {
        if (!this._ami)
            throw new Error('no valid ami');
        return this._ami;
    }

    /**
     * @param config    is configuration to make AstUtils
     */
    constructor(config: AstUtilsConfg) {
        this._ari = new Ari({
            url: `${config.ari.protocol}://${config.host}:${config.ari.port}`,
            user: config.ari.username,
            password: config.ari.password
        });
        this._ami = new Ami({
            host: config.host,
            port: config.ami.port,
            user: config.ami.username,
            secret: config.ami.password
        });
    }

    /**
     * connect a remort asterisk
     */
    async connect(): Promise<void> {
        await this.ari.connect();
        await this.ami.connect();
    }

    /**
     * disconnect from the remort asterisk
     */
    disconnect(): void {
        this.ari.disconnect();
        this.ami.disconnect();
    }

    /**
     * get an endpoint status of the remote asterisk
     * @param id    is id for the endpoint you want
     */
    getEndpointStatus(id: string): Promise<string> {
        return this.ari.getEndpointStatus(id);
    }

    /**
     * exec the specified command
     */
    exec(command: string): Promise<AstUtilsResponse> {
        debug('exec', command);
        return this.ami.exec(command);
    }

    /**
     * let a remote asterisk reload dial plan
     */
    reloadDialPlan(): Promise<AstUtilsResponse> {
        debug('reloadDialPlan');
        return this.ami.reloadDialPlan();
    }

    /**
     * execute 'core reload'
     * @param timer time to wait after the command has been executed, default is 1000msec.
     * @return  when the spcified timer has been expired
     *          after 'core reload' command is executed.
     */
    async reload(timer = 1000): Promise<void> {
        debug('reload', timer);
        await this.ami.reload();
        return new Promise<void>(resolve => setTimeout(resolve, timer));
    }

    /**
     * let a remote asterisk-self restart
     * @param timer time to wait after reconnection, default is 1000msec.
     * @return  when the spcified timer has been expired
     *          after a connection is reestablished.
     */
    async restart(timer = 1000): Promise<void> {
        debug('restart', timer);
        await this.ami.restart();
        return new Promise<void>(resolve => setTimeout(resolve, timer));
    }

    /**
     * let a remote server call to callee with specified sound
     * @param callee    id of callee
     * @param sound     sound file to play while calling
     */
    callWithSound(callee: string, sound: string): Promise<AstUtilsResponse> {
        debug('callWithSound', callee, sound);
        return this.ami.callWithSound(callee, sound);
    }
}
