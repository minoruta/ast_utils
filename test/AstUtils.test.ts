import { AstUtils } from '../src/AstUtils';

let asterisk: AstUtils;

describe('AstUtils.utils', () => {
    test ('connect to asterisk', async () => {
        asterisk = new AstUtils({
            host: '127.0.0.1',
            ari: {
                protocol: 'http',
                port: 8088,
                username: 'asterisk',
                password: 'asterisk'
            },
            ami: {
                port: 5038,
                username: 'asterisk',
                password: 'asterisk'
            }
        });
        await asterisk.connect();
    });
    test ('reload dialplan', async () => {
        const response = await asterisk.reloadDialPlan();
        expect(response).toMatchSnapshot();
    });

    test ("get status of an unknown endpoint", async () => {
        const result = await asterisk.getEndpointStatus('1234567890');
        expect(result).toBe('no endpoint');
    });

    test ("get status of a valid endpoint", async () => {
        const result = await asterisk.getEndpointStatus('6001');
        expect(result).toBe('offline');
    });

    test ("get list of a endpoints", async () => {
        const result = await asterisk.exec('core show codecs');
        expect(result).toMatchSnapshot();
    });

    test ("call with sound", async () => {
        const response = await asterisk.callWithSound('6001', 'hello');
        expect(response).toMatchSnapshot();
    });

    test ('restart asterisk', async () => {
        await asterisk.restart();
    });

    test ('disconnet from the remote asterisk', async () => {
        await asterisk.disconnect();
    });
});
