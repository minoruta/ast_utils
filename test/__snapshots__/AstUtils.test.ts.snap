// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstUtils.utils call with sound 1`] = `
Object {
  "Message": "Command output follows",
  "Output": "",
  "Response": "Success",
}
`;

exports[`AstUtils.utils get list of a endpoints 1`] = `
Object {
  "Message": "Command output follows",
  "Output": "Disclaimer: this command is for informational purposes only.
It does not indicate anything about your configuration.
ID TYPE  NAME         FORMAT           DESCRIPTION
------------------------------------------------------------------------------------------------
31 image png          png              (PNG Image)
6 audio g726         g726             (G.726 RFC3551)
4 audio alaw         alaw             (G.711 a-law)
2 audio g723         g723             (G.723.1)
20 audio speex        speex            (SpeeX)
21 audio speex        speex16          (SpeeX 16khz)
22 audio speex        speex32          (SpeeX 32khz)
24 audio g722         g722             (G722)
25 audio siren7       siren7           (ITU G.722.1 (Siren7, licensed from Polycom))
32 video h261         h261             (H.261 video)
33 video h263         h263             (H.263 video)
8 audio adpcm        adpcm            (Dialogic ADPCM)
43 audio silk         silk8            (SILK Codec (8 KHz))
44 audio silk         silk12           (SILK Codec (12 KHz))
45 audio silk         silk16           (SILK Codec (16 KHz))
46 audio silk         silk24           (SILK Codec (24 KHz))
28 audio g719         g719             (ITU G.719)
34 video h263p        h263p            (H.263+ video)
35 video h264         h264             (H.264 video)
19 audio g729         g729             (G.729A)
9 audio slin         slin             (16 bit Signed Linear PCM)
10 audio slin         slin12           (16 bit Signed Linear PCM (12kHz))
11 audio slin         slin16           (16 bit Signed Linear PCM (16kHz))
12 audio slin         slin24           (16 bit Signed Linear PCM (24kHz))
13 audio slin         slin32           (16 bit Signed Linear PCM (32kHz))
14 audio slin         slin44           (16 bit Signed Linear PCM (44kHz))
15 audio slin         slin48           (16 bit Signed Linear PCM (48kHz))
16 audio slin         slin96           (16 bit Signed Linear PCM (96kHz))
17 audio slin         slin192          (16 bit Signed Linear PCM (192kHz))
3 audio ulaw         ulaw             (G.711 u-law)
18 audio lpc10        lpc10            (LPC10)
27 audio testlaw      testlaw          (G.711 test-law)
42 audio none         none             (<Null> codec)
41 image t38          t38              (T.38 UDPTL Fax)
38 video vp9          vp9              (VP9 video)
37 video vp8          vp8              (VP8 video)
5 audio gsm          gsm              (GSM)
36 video mpeg4        mpeg4            (MPEG4 video)
23 audio ilbc         ilbc             (iLBC)
39 text  red          red              (T.140 Realtime Text with redundancy)
40 text  t140         t140             (Passthrough T.140 Realtime Text)
29 audio opus         opus             (Opus Codec)
30 image jpeg         jpeg             (JPEG image)
7 audio g726aal2     g726aal2         (G.726 AAL2)
1 audio codec2       codec2           (Codec 2)
26 audio siren14      siren14          (ITU G.722.1 Annex C, (Siren14, licensed from Polycom))",
  "Response": "Success",
}
`;

exports[`AstUtils.utils reload dialplan 1`] = `
Object {
  "Message": "Command output follows",
  "Output": "Dialplan reloaded.",
  "Response": "Success",
}
`;
